core/src/main/scala/caliban/parsing/Parser.scala

```diff
@@ -266,6 +267,161 @@ object Parser {
 
+  private def schemaExtensionWithDirectivesAndOptionalOperations[_: P]: P[SchemaExtension] =
+    P(directives ~ ("{" ~ rootOperationTypeDefinition.rep ~ "}").?).map {
+      case (directives, Some(ops)) =>
+        val opsMap = ops.toMap
+        SchemaExtension(
+          directives,
+          opsMap.get(OperationType.Query).map(_.name),
+          opsMap.get(OperationType.Mutation).map(_.name),
+          opsMap.get(OperationType.Subscription).map(_.name)
+        )
+      case (directives, None) =>
+        SchemaExtension(
+          directives,
+          None,
+          None,
+          None
+        )
+    }
+
+  private def schemaExtensionWithOperations[_: P]: P[SchemaExtension] =
+    P("{" ~ rootOperationTypeDefinition.rep ~ "}").map { ops =>
+      val opsMap = ops.toMap
+      SchemaExtension(
+        Nil,
+        opsMap.get(OperationType.Query).map(_.name),
+        opsMap.get(OperationType.Mutation).map(_.name),
+        opsMap.get(OperationType.Subscription).map(_.name)
+      )
+    }
+
+  private def schemaExtension[_: P]: P[SchemaExtension] =
+    P("extend schema" ~/ (schemaExtensionWithDirectivesAndOptionalOperations | schemaExtensionWithOperations))
+
+  private def scalarTypeExtension[_: P]: P[ScalarTypeExtension] =
+    P("extend scalar" ~/ name ~ directives).map {
+      case (name, directives) =>
+        ScalarTypeExtension(name, directives)
+    }
+
+  private def objectTypeExtensionWithInterfacesAndOptionalFields[_: P]: P[ObjectTypeExtension] =
+    P(name ~ implements ~ !directives ~ ("{" ~ fieldDefinition.rep ~ "}").?).map {
+      case (name, implements, fields) =>
+        ObjectTypeExtension(
+          name,
+          implements,
+          Nil,
+          fields.map(_.toList).getOrElse(Nil)
+        )
+    }
+
+  private def objectTypeExtensionWithInterfacesDirectivesAndOptionalFields[_: P]: P[ObjectTypeExtension] =
+    P(name ~ implements ~ directives ~ ("{" ~ fieldDefinition.rep ~ "}").?).map {
+      case (name, implements, directives, fields) =>
+        ObjectTypeExtension(
+          name,
+          implements,
+          directives,
+          fields.map(_.toList).getOrElse(Nil)
+        )
+    }
+
+  private def objectTypeExtensionWithDirectivesAndOptionalFields[_: P]: P[ObjectTypeExtension] =
+    P(name ~ directives ~ ("{" ~ fieldDefinition.rep ~ "}").?).map {
+      case (name, directives, fields) =>
+        ObjectTypeExtension(
+          name,
+          Nil,
+          directives,
+          fields.map(_.toList).getOrElse(Nil)
+        )
+    }
+
+  private def objectTypeExtensionWithFields[_: P]: P[ObjectTypeExtension] =
+    P(name ~ "{" ~ fieldDefinition.rep ~ "}").map {
+      case (name, fields) =>
+        ObjectTypeExtension(
+          name,
+          Nil,
+          Nil,
+          fields.toList
+        )
+    }
+
+  private def objectTypeExtension[_: P]: P[ObjectTypeExtension] =
+    P(
+      "extend type" ~/ (
+        objectTypeExtensionWithInterfacesAndOptionalFields |
```

Comment:
I think you can reduce to 3 cases:
- fields + optional interfaces + optional directives
- no fields + directives + optional interfaces
- no fields, no directives, interfaces

Reply:
@ghostdogpr Yeah, actually my first attempt was modelling the Parser more strictly after the specs, like this:
```scala
private def objectTypeExtensionWithOptionalInterfacesOptionalDirectivesAndFields[_: P]: P[ObjectTypeExtension] =
    P(name ~ implements.? ~ directives.? ~ "{" ~ fieldDefinition.rep ~ "}").map {
      case (name, implements, directives, fields) =>
        ObjectTypeExtension(
          name,
          implements.getOrElse(Nil),
          directives.getOrElse(Nil),
          fields.toList
        )
    }

  private def objectTypeExtensionWithOptionalInterfacesAndDirectives[_: P]: P[ObjectTypeExtension] =
    P(name ~ implements.? ~ directives).map {
      case (name, implements, directives) =>
        ObjectTypeExtension(
          name,
          implements.getOrElse(Nil),
          directives,
          Nil
        )
    }

  private def objectTypeExtensionWithInterfaces[_: P]: P[ObjectTypeExtension] =
    P(name ~ implements).map {
      case (name, implements) =>
        ObjectTypeExtension(
          name,
          implements,
          Nil,
          Nil
        )
    }


  private def objectTypeExtension[_: P]: P[ObjectTypeExtension] =
    P(
      "extend type" ~/ (
        objectTypeExtensionWithOptionalInterfacesOptionalDirectivesAndFields |
          objectTypeExtensionWithOptionalInterfacesAndDirectives |
          objectTypeExtensionWithInterfaces
      )
    )
```

Reply:
However, that didn't work as expected, when running unit tests I've got failures for some cases. Here are the error messages I got (I've enabled verboseFailures temporally):

Reply:
```
 extend type with interfaces and directives
[info]     Fiber failed.
[info]     A checked error was not handled.
[info]     Parsing Error: Expected document:1:1 / objectTypeExtension:1:1 / objectTypeExtensionWithOptionalInterfacesOptionalDirectivesAndFields:1:13 / "{":1:58, found "" 
```

Reply:
```
extend type with directives
[info]     Fiber failed.
[info]     A checked error was not handled.
[info]     Parsing Error: Expected document:1:1 / objectTypeExtension:1:1 / objectTypeExtensionWithOptionalInterfacesOptionalDirectivesAndFields:1:13 / "{":1:33, found "" 
```

Reply:
I've tried other combinations as well, but none of them worked for all cases, that's why I had to implement the parser rules in a different way.

Reply:
What do you think @ghostdogpr? Maybe I'm missing something?

Reply:
Hmm that’s strange indeed. I can’t see anything wrong with the code at first sight but tomorrow I’ll get your branch and try to figure it out in debug.

Reply:
And, I got similar issues when I tried to define parsers for the other type extension cases

Reply:
Awesome, thanks for your help @ghostdogpr!